### ISSUE TITLE
Datapoint reports rawdata, needs to be energy information (phase_a_ + current, voltage, power)

### ISSUE THREAD

**Issue by BartGysens** [rospogrigio/localtuya#1193]
## The problem

Device: TUYA APP 1P 40A Rail Din WIFI compteur d'énergie intelligent consommation d'énergie kWh interrupteur de disjoncteur
(link: [https://www.amazon.fr/dp/B09FPGNYFK/ref=pe_27091421_487030221_TE_SCE_dp_1])

When I try to get data in localtuya, the power consumption (which works fine in Tuya-cloud) comes in as rawdata.
I think the string is a Base64-coded value, but I do not known how to convert this to current, voltage and power.

This is the reading I get on datapoint 6 (it comes as rawdata): CWYAAAAAAAA= x
This DP is described in the Tuya Cloud as "phase_a".
Should this value be converted into an array (because I saw this in other devices)? How?

Can anybody help pls? Thank you.

## Environment
- Localtuya version: 4.1.1 
- Home Assistant Core version: Home Assistant OS 9.4, version: 2022.12.8
- [x] Are you using the Home Assistant Tuya Cloud component ?
- [x] Are you using the Tuya App in parallel ?

## Configuration `configuration.yaml` or `config_flow`
`"bf69893b9f9e8b7ee08vuu": {
        "friendly_name": "Tuya disjoncteur/compteur DIN 40A",
        "local_key": "c36...a09",
        "host": "192.168.0.222",
        "device_id": "bf69893b9f9e8b7ee08vuu",
        "protocol_version": "3.3",
        "manual_dps_strings": "6",
        "model": "breaker",
        "dps_strings": [
          "1 (value: 6747)",
          "9 (value: 0)",
          "11 (value: False)",
          "12 (value: False)",
          "13 (value: 0)",
          "14 (value: 0)",
          "15 (value: 0)",
          "16 (value: True)",
          "19 (value: 4567325321)",
          "21 (value: False)",
          "6 (value: -1)"
        ],
        "entities": [
          {
            "id": 16,
            "friendly_name": "Boiler (Lavande)",
            "restore_on_reconnect": false,
            "is_passive_entity": false,
            "platform": "switch"
          },
          {
            "id": 1,
            "friendly_name": "Boiler (Lavande) - energieverbruik",
            "unit_of_measurement": "kWh",
            "device_class": "power",
            "scaling": 0.01,
            "platform": "sensor"
          },
          {
            "id": 6,
            "friendly_name": "Boiler (Lavande) - Base64-encoded raw data",
            "unit_of_measurement": "x",
            "device_class": "apparent_power",
            "scaling": 1.0,
            "platform": "sensor"
          }
        ],
        "product_key": "hz2o2tnezttmmx15"
`

## DP dump
I added the DP6 manually while adding the device in localtuya because it wasn't detected automaticly.

`
total_forward_energy | Integer | {   "unit": "kW·h",   "min": 0,   "max": 99999999,   "scale": 2,   "step": 1 }
-- | -- | --
phase_a | Raw | {}
fault | Bitmap | {   "label": [     "short_circuit_alarm",     "surge_alarm",     "overload_alarm",     "leakagecurr_alarm",     "temp_dif_fault",     "fire_alarm",     "high_power_alarm",     "self_test_alarm",     "ov_cr",     "unbalance_alarm",     "ov_vol",     "undervoltage_alarm",     "miss_phase_alarm",     "outage_alarm",     "magnetism_alarm",     "credit_alarm",     "no_balance_alarm"   ] }
switch_prepayment | Boolean | "{true,false}"
energy_reset | Enum | {   "range": [     "empty"   ] }
balance_energy | Integer | {   "unit": "kW·h",   "min": 0,   "max": 99999999,   "scale": 2,   "step": 1 }
charge_energy | Integer | {   "unit": "kW·h",   "min": 0,   "max": 999999,   "scale": 2,   "step": 1 }
leakage_current | Integer | {   "unit": "mA",   "min": 0,   "max": 1000,   "scale": 0,   "step": 1 }
switch | Boolean | "{true,false}"
alarm_set_1 | Raw | {}
alarm_set_2 | Raw | {}
breaker_number | String | {   "maxlen": 255 }
leakagecurr_test | Boolean | "{true,false}"
`

**Comment by MaximumSU** [comment]
Have the same issue with 3 phase socket breaker with meetering capability. It report Voltage, Current and another values of 3 phases in one DP as base64 encoded bites (2 byte+2byte+2byte for each A,B,C phase). Your decoded value is 240.6 0 0 by this method. Theese DP is not reported by device itself nor in tuya iot API. Sadly, this functionality (DP values binary decoder) is not realised yet in localtuya.

**Comment by BartGysens** [comment]
@MaximumSU how did you decode this string 'CWYAAAAAAAA=' , I tried base64_decode in python but it didn't give me any values, only an x,f,m,...

**Comment by MaximumSU** [comment]
@BartGysens 
There are 4 2xByte values in your payload. Decode all string with base64 then take first (second, third, fourth) two bytes and convert it with int.from_bytes(state, byteorder='big'), then multiple by 0.1 - you taked the final value
I created the fork of this project and added this functionality [here](https://github.com/MaximumSU/localtuya)

**Comment by BartGysens** [comment]
@MaximumSU thank you for talking me through this!

This is my code and something is not right...
It gives me a Voltage = 1388498.3 where it should be 236

`@state_trigger('sensor.boiler_lavande_phase_a')
def base64_decode_sensor_lavande():
    #!/usr/bin/env python3
    import base64
    import sys

    value_phase_a = state.get('sensor.boiler_lavande_phase_a')
    #state.set('sensor.boiler_lavande_voltage_script', value_phase_a)

    value_decoded = base64.b64decode(value_phase_a).hex()
    #state.set('sensor.boiler_lavande_voltage_script', value_decoded[0:4])

    value_voltage = base64.b64decode(value_decoded[0:4])
    #state.set('sensor.boiler_lavande_voltage_script', value_voltage)

    value_voltage = int.from_bytes( base64.b64decode(value_decoded[0:4]), 'big')*0.1
    state.set('sensor.boiler_lavande_voltage_script', value_voltage)`

**Comment by MaximumSU** [comment]
@BartGysens ".hex" not need (it produces hexadecimal string, but you need the binary string of bytes, it is already here) and 2 bytes ([0:2]), not 3 ([0:4])
See below: 
```
value_phase_a = state.get('sensor.boiler_lavande_phase_a')
value_voltage = round(int.from_bytes(base64.b64decode(value_phase_a)[0:2], byteorder='big')*0.1)
state.set('sensor.boiler_lavande_voltage_script', value_voltage)

```

**Comment by DominikWrobel** [comment]
Hi, I've used your solution for my DDS238-2 meter made by OXT, but my data points were different, here is the code:

`@state_trigger(‘sensor.licznik_pradu_dol_fazaa’)
def base64_decode_sensor_licznik():
#!/usr/bin/env python3
import base64
import sys

# Decoding value from datapoint 6 which is base64-based (manually activated)
value_phase_a = state.get('sensor.licznik_pradu_dol_fazaa')
value_decoded = base64.b64decode(value_phase_a)

# Calculating Voltage from base64-decoded value
value_voltage = round(int.from_bytes(value_decoded[13:15], byteorder='big')*0.1, 3)
state.set('sensor.licznik_pradu_dol_voltage_py', value_voltage)

# Calculating Current from base64-decoded value
value_current = round(int.from_bytes(value_decoded[11:13], byteorder='big')*0.001, 3)
state.set('sensor.licznik_pradu_dol_current_py', value_current)

# Calculating Power from base64-decoded value
value_power = round(int.from_bytes(value_decoded[2:4], byteorder='big')*0.001, 3)
state.set('sensor.licznik_pradu_dol_power_py', value_power)

# Calculating Reactive Power from base64-decoded value
value_kvar = round(int.from_bytes(value_decoded[6:8], byteorder='big')*0.001, 3)
state.set('sensor.licznik_pradu_dol_kvar_py', value_kvar)

# Calculating Reactive PF from base64-decoded value
value_pf = round(int.from_bytes(value_decoded[8:10], byteorder='big')*0.001, 3)
state.set('sensor.licznik_pradu_dol_pf_py', value_pf)`

Now I have one problem, for my power value in the app I get positive power when getting power from the grid and a negative value when my solar is sending power to grid, but the value from the script is always positive. How can I correct that?

**Comment by BartGysens** [comment]
I did some Google; maybe this could help you further:

https://stackoverflow.com/questions/48885704/decode-base64-encoded-byte-array-to-negative-decimal-value-java-to-python

```coded_string = "/bfbKNk="
bytes_val = base64.decodesbytes(coded_string.encode())
bval = "".join("{0:08b}".format(c) for c in bytes_val)
intval = int(bval, 2)
if bytes_val[0] & 0x70 != 0:           # manually takes the two's complement
    intval -= int('1' + '00' * len(bytes_val), 16)
print(intval)```

**Comment by MaximumSU** [comment]
> but the value from the script is always positive

So, value is correct (except of sign), but positive? Or it is positive and incorrect?

**Comment by DominikWrobel** [comment]
The value is correct! But when the sun is out and my solar is working in the Tuya app I see the same value as in the py script but with a negative sign.  When solar is not working everything is ok.

**Comment by MaximumSU** [comment]
I think, it is one more bit/byte/DP especially for the direction. Need to find :-)
I see you used the bytes [2:4],[6:8],[8:10],[11:13],[13:15]
What is the values in bytes 0,1,4,5? It is possible the sign of direction there..

**Comment by DominikWrobel** [comment]
Ok, so I think it is the [0:1] bit! it gives me AA (0) or AQ (1) value and when checking the history on HA and on the tuya app I see that the value changed from AA to AQ when the solar started working and my power consumption from grid fell to 0. Now can you help me get this to the python code? I could check for a couple of days to see if it works ok.

**Comment by MaximumSU** [comment]
@DominikWrobel Give me please the full raw (encoded, as received from device) data strings with positive and negative values.

**Comment by DominikWrobel** [comment]
Positive values:  AAABzQAAADIDdgAIywkH
Negative: AQAErgAAALoD1QAUPgkz

**Comment by MaximumSU** [comment]
@DominikWrobel Something like this:
`value_power = round(int.from_bytes(value_decoded[2:4], byteorder='big')*0.001, 3) * (-1 if ord(value_decoded[0:1]) else 1)`

**Comment by DominikWrobel** [comment]
@MaximumSU thank you - I think this will do the trick. I will check the resoults next couple of days and let you know if it works as it should.

**Comment by mwkchan** [comment]
@MaximumSU I am using your fork. How should I fill in the Binary Segmentation in your localtuya UI? It always my format is wrong. I should be using '0:2';'3:5';'6:8'. The only time I got it success is to just enter 0:2 and it gave me the voltage.

**Comment by MaximumSU** [comment]
@mwkchan 
> I am using your fork. How should I fill in the Binary Segmentation in your localtuya UI? It always my format is wrong. I should be using '0:2';'3:5';'6:8'. The only time I got it success is to just enter 0:2 and it gave me the voltage.

1. Just to confirm: are you use 3 different sensors? You not write '0:2';'3:5';'6:8' as segmentation rule in one sensor? You shoud have 3 different sensors...
2. You shoud not use the positions in the two arguments, but the beginning and count of bytes. For example - not '3:5' but '3:2'. '3:5' means '5 bytes from position 3' that is wrong.

**Comment by MaximumSU** [comment]
> Are you going to ask the main repo in HACS incorporating your contributions?

Yes, i do. It is here #1221
But it is not accepted yet.

**Comment by BaltasarParreira** [comment]
@MaximumSU I'm using your fork but having some problems on one of the DP, this one:

```
      {
        "code": "lunar_cycle",
        "custom_name": "",
        "dp_id": 102,
        "time": 1696591156422,
        "value": "ARY="
      },
```

It has 2 bytes that decoding gives this:
01 16

meaning 01 is on (oo was off) and the 2nd value converting 16 hex to decimal is 22, the lunar calendar day number.

But the real problem is that I add this DP and allways get this error:
![image](https://github.com/rospogrigio/localtuya/assets/1650425/68b977d3-5911-43d1-99df-3e03ded361ae)

and can't find any solution.

The ID is added like this (I need only the second byte for the moment):
![image](https://github.com/rospogrigio/localtuya/assets/1650425/4b414c45-08ba-416e-a381-63ee14af45c0)

Any tip or solution for this problem?
Thanks

**Comment by BaltasarParreira** [comment]
You mean this (exert from core.config_entries file):

```
            "bffad3aa56e6417a01mbsa": {                               
              "friendly_name": "Aqua Reef Switch All",                
              "host": "192.168.0.150",                                
              "local_key": "xxxxxxxxxxxxx",                        
              "protocol_version": "3.3",                              
              "enable_debug": true,                                   
              "manual_dps_strings": "101,102,103,106,107,108,109,110",
              "reset_dpids": "",                                   
              "entities": [                                           
                {                                                     
                  "friendly_name": "Aqua Reef Switch All",            
                  "restore_on_reconnect": false,                      
                  "is_passive_entity": false,                         
                  "id": 101,                                          
                  "platform": "switch"                         
                },                                             
                {                                              
                  "friendly_name": "Aqua Reef Mode",           
                  "select_options": "curve;lighting;fast;show",
                  "restore_on_reconnect": false,               
                  "is_passive_entity": false,                  
                  "id": 103,                                   
                  "platform": "select"                         
                },                                             
                {                                              
                  "friendly_name": "Aqua Reef Current Time",   
                  "min_value": 0.0,                         
                  "max_value": 100000.0,                    
                  "step_size": 1.0, 
                 "restore_on_reconnect": false,                      
                  "is_passive_entity": false,                         
                  "id": 110,                                          
                  "platform": "number"                                
                },                                                    
                {                                                     
                  "friendly_name": "Aqua Reef Ch Data",               
                  "base64_decode": true,                              
                  "bytes_range": "3:1",                               
                  "id": 108,                                          
                  "platform": "sensor"                                
                },                                                    
                {                                                     
                  "friendly_name": "test",                            
                  "base64_decode": true,                              
                  "bytes_range": "1:1",                               
                  "id": 108,                                   
                  "platform": "sensor"                         
                },                                             
                {                                              
                  "friendly_name": "Aqua Reef Lunar Cycle",    
                  "base64_decode": true,                       
                  "bytes_range": "1:1",                        
                  "scaling": 1.0,                              
                  "id": 102,                                   
                  "platform": "sensor"                         
                } 
             ],                                                      
              "add_entities": false,                           
              "device_id": "bffad3aa56e6417a01mbsa",           
              "dps_strings": [                                 
                "101 (value: True)",                           
                "103 (value: curve)",                          
                "110 (value: 6)",                              
                "102 (value: -1)",                             
                "106 (value: -1)",                             
                "107 (value: -1)",                             
                "108 (value: -1)",                             
                "109 (value: -1)"                              
              ],                                               
              "product_key": "xxxxxxxxxxxxx"             
            } 
```

The only thing I can see in the log regarding DP 102 is this:

`2023-10-06 13:23:01.153 WARNING (SyncWorker_8) [custom_components.localtuya.sensor] [bff...bsa] Entity sensor.aqua_reef_lunar_cycle is requesting unknown DPS index 102`

Changing the scaling doesn't make any difference.

**Comment by MaximumSU** [comment]
@BaltasarParreira Are you sure your device have DP102 ?

**Comment by BaltasarParreira** [comment]
@MaximumSU  yes, the tuya dev site shows it and the values changed when I change things for "unar cycle" in the Tuya app on the phone.

Here is the DPs checking on tuya website (Parameter(Request Method: GET)):
```
{
  "result": {
    "properties": [
      {
        "code": "switch_all",
        "custom_name": "",
        "dp_id": 101,
        "time": 1696454949924,
        "value": true
      },
      {
        "code": "lunar_cycle",
        "custom_name": "",
        "dp_id": 102,
        "time": 1696591156422,
        "value": "ARY="
      },
      {
        "code": "mode",
        "custom_name": "",
        "dp_id": 103,
        "time": 1696590561024,
        "value": "curve"
      },
      {
        "code": "accalimation",
        "custom_name": "",
        "dp_id": 106,
        "time": 1696454950122,
        "value": "AAECZA=="
      },
      {
        "code": "edit_curve",
        "custom_name": "",
        "dp_id": 107,
        "time": 1696454950227,
        "value": "FBAAADIAAAQAMgAABwAAAABAAAAAAEIACgAFRAAZAApGBS0FD0gPUBkUSh5aMiNMMmRkPFcyZGQ8WR5aMiNbFFAZFF0FMgUKXgAyAAVfADIAAA=="
      },
      {
        "code": "Ch_data",
        "custom_name": "",
        "dp_id": 108,
        "time": 1696597980708,
        "value": "AAAAAA=="
      },
      {
        "code": "ch_control",
        "custom_name": "",
        "dp_id": 109,
        "time": 1693939181193
      },
      {
        "code": "current_time",
        "custom_name": "",
        "dp_id": 110,
        "time": 1696597201544,
        "value": 56
      }
    ]
  },
  "success": true,
  "t": 1696597991048,
  "tid": "1948aaf4644a11eebb8942dfc87a0fc7"
}
```

**Comment by BaltasarParreira** [comment]
@MaximumSU I just finish test all DPs that have raw value and the only one that works is the 108 that I already have 2 setup, all the others not working.
I mean 102, 106, 107 and 109 that I don't know what is for since doesn't have value and nothing changes when I change anything on the Tuya phone app).

**Comment by MaximumSU** [comment]
@BaltasarParreira Are you tried "DPIDs to send in RESET command (separated by commas ',')- Used when device does not respond to status requests after turning on (optional)" in device settings?

**Comment by BaltasarParreira** [comment]
@MaximumSU No, still no go.
I put all DPs that are not working on the reset but none of those are working.

![image](https://github.com/rospogrigio/localtuya/assets/1650425/2fb73fa0-ce31-4cfc-898f-3cccb1af8d99)

Should I put on the manual DPs the ones that are already auto detected or just the extra ones?
As you can see on the image my setup is like that now.

**Comment by MaximumSU** [comment]
@BaltasarParreira What is the name of your device? What is the HASS entity name of your device? Please [enable debug](https://github.com/rospogrigio/localtuya#debugging) for your device and restart Home Assistant. I need to see the startup log.

PS:Your issue is not related to my modifications (data from device just not come to HASS, there are may be many reasons for that), it is somethimg else. But i try to help you anyway :-)

**Comment by BaltasarParreira** [comment]
@MaximumSU Yeahh is not your software as the any other tuya soft I get the same problem, but it's strange the 108 working ok and all the others not, anyway this is the main entity for my device (circle in red):

![image](https://github.com/rospogrigio/localtuya/assets/1650425/2305c55c-ec1a-4eff-9a88-1bf66360379c)

And the name on HASS is: switch.aqua_reef_switch_all

Regarding the debug log you mean turning this on?

![image](https://github.com/rospogrigio/localtuya/assets/1650425/056e03b7-abaf-42c2-95e5-d5d6ddbf58a5)

**Comment by BaltasarParreira** [comment]
@MaximumSU I think the log you want is the file that saves when after rebooting I clicked the "Disable debug logging" button on local tuya, right?

Here is that file:
[home-assistant_localtuya_2023-10-07T09-47-53.845Z.log](https://github.com/rospogrigio/localtuya/files/12837381/home-assistant_localtuya_2023-10-07T09-47-53.845Z.log)

**Comment by MaximumSU** [comment]
@BaltasarParreira There are nothing useful in logs. I think, your device is one of devices which present DPID of the main switch, but not the additional sensors DPIDs. For them reset is needed before status. This acheve by RESET DPID's, but this logic not implemented yet. I can try to do some modification of the code, but i can't test it.

**Comment by BaltasarParreira** [comment]
@MaximumSU If you tell me what I need to change I can do it and test it.
Or you can send your changes so I will test then.
Whatever is better for you I'm here to help on all tests needed.
Thanks again for everything in trying to help solving this.

**Comment by MaximumSU** [comment]
@BaltasarParreira Replace this files in custom_components/localtuya folder (backup old files first!).
[localtuya.zip](https://github.com/rospogrigio/localtuya/files/12838522/localtuya.zip)
then restart Home Assistant. The required DPS must be mentioned in "DPIDs to send in RESET command"

**Comment by BaltasarParreira** [comment]
@MaximumSU I think there is some error on the new code somewhere, as I get this message:
`2023-10-07 20:07:57.375 DEBUG (MainThread) [custom_components.localtuya.pytuya] [bff...bsa] ERROR Unexpected Payload from Device - 904 - payload: null`

I think after sending the reset you said:
`2023-10-07 20:07:56.841 DEBUG (MainThread) [custom_components.localtuya.pytuya] [bff...bsa] Sending payload: b'{"dpId":[102,106,107,108,109]}'`

But here the new full debug log:
[home-assistant_localtuya_2023-10-07T19-09-26.114Z.log](https://github.com/rospogrigio/localtuya/files/12838591/home-assistant_localtuya_2023-10-07T19-09-26.114Z.log)

**Comment by BaltasarParreira** [comment]
@MaximumSU Looking to some of the code differences looks like it didn't reach this important part from common.py file:
```
self.debug(
                        "Resetting command for DP IDs: %s",
                        self._default_reset_dpids,
                    )
```

as it can't be seeing on the log file.

**Comment by MaximumSU** [comment]
@BaltasarParreira May be the Tuya IOT web site have some clue in Device Log section?

**Comment by BaltasarParreira** [comment]
@MaximumSU Hummm... don't really know what you need?

Can you explain or detail it?
From here or somewhere else any of this tests /diagnostics to run?

![image](https://github.com/rospogrigio/localtuya/assets/1650425/ba2936dc-9223-474f-b476-5fa38b19ccce)

**Comment by MaximumSU** [comment]
@BaltasarParreira 
Cloud -> Open Project -> Devices -> Debug Device -> Device Logs
There are DP ID's and events from/to device on that page. May be we can see there some clue, what the application do.

**Comment by BaltasarParreira** [comment]
@MaximumSU Strangely it looks that tuya is treating the device as a normal lamp and It's not, anyway here is all the info that shows for the week (sorry it can't be exported as text so I have to put here as images):
![image](https://github.com/rospogrigio/localtuya/assets/1650425/02670bd4-d270-4f13-af87-56ec28d3e49e)

![image](https://github.com/rospogrigio/localtuya/assets/1650425/8c0e96d5-38d3-4691-a3c5-d8a1a8fec9be)

![image](https://github.com/rospogrigio/localtuya/assets/1650425/2391c5db-aff3-4287-9b7d-b39e5c487193)

![image](https://github.com/rospogrigio/localtuya/assets/1650425/a6bead42-4ddc-4832-a6bc-15dc33e57967)

**Comment by MaximumSU** [comment]
@BaltasarParreira No, not this. This is the report of device, need something just before the report. Something, what sended to device to get report...

**Comment by BaltasarParreira** [comment]
@MaximumSU but that log shows that?
I only see value reports nothing else, apart once in a while this:

![image](https://github.com/rospogrigio/localtuya/assets/1650425/6a3d304b-bd55-4e84-ba24-a1d4472b9527)

**Comment by MaximumSU** [comment]
@BaltasarParreira This is some statistic, does not matter.
On the first screenshot i see the time 10:27:44 and 10:27:52
Second screen showing another times. Need all events and DPS, as from second screen, but at the time of the first screen.

**Comment by BaltasarParreira** [comment]
@MaximumSU Well I think there is nothing sort of that information showing, but here is all I get for a near by time interval:
![image](https://github.com/rospogrigio/localtuya/assets/1650425/50afcfb9-f275-4adf-ad9a-41ab84945400)

![image](https://github.com/rospogrigio/localtuya/assets/1650425/89020795-7321-4cfe-bbba-9722f9341426)

![image](https://github.com/rospogrigio/localtuya/assets/1650425/1938ac95-009d-483f-b3c8-dc640f872916)

![image](https://github.com/rospogrigio/localtuya/assets/1650425/be8167b8-5124-4c20-9599-ee747cf854e4)

**Comment by BaltasarParreira** [comment]
@MaximumSU Should I change this, default is "Standard Instruction", right?

![image](https://github.com/rospogrigio/localtuya/assets/1650425/94ca80c7-48e9-4d3b-9d98-1ee6615ea61f)

or is nothing related with this?
I'm in the dark here and don't know what else to do....    :  (

**Comment by MaximumSU** [comment]
@BaltasarParreira I have no more ideas. Sorry, but without having device and no full access to tuya iot data farfer investigation is not possible.

**Comment by BaltasarParreira** [comment]
@MaximumSU  Yes I understand, access to tuya iot can be solved, but to the device itself will be difficult or really impossible.
Anyway if it helps some sort of call trough Skype, Teams and sharing screens or using some other remote app if you have the time to explore this further it's just a matter of try to arrange something, but it's up with you off course.

**Comment by BaltasarParreira** [comment]
@MaximumSU I was looking all around the HASS interface and apart the entity for DP 102 is always saying unavaible:

![image](https://github.com/rospogrigio/localtuya/assets/1650425/ae4fa62d-9854-4859-8f87-331c52e511eb)

I see one value there (23 for the 23th lunar day of moon calendar, was yesterday) so I went to history and some times there is more values for DP 102, like for today at some time 24th moon day, this is strange:

![image](https://github.com/rospogrigio/localtuya/assets/1650425/21de383b-2233-423d-a6e8-35319e05de28)

**Comment by MaximumSU** [comment]
@BaltasarParreira For example, It may works then App on mobile is open.

**Comment by BaltasarParreira** [comment]
@MaximumSU You are correct, when I open de tuya phone app and if I go to the moon cycle menu and turn it on them off or vice versa the HASS starts working and no more unavailable.
Going to leave it like this and see if tomorrow the moon day changes auto for next day without touching the phone app.

Anyway after doing this I look again on the tuya iot device debug log to see if anything different shows, but apart the normal value passing because of the changing on/off just some new information for other DP, like this:
![image](https://github.com/rospogrigio/localtuya/assets/1650425/809a90db-12d4-4740-bb7c-90d6a4918210)

"Curve" value is from DP 103.

**Comment by fjramos80** [comment]
#!/usr/bin/env python3
    import base64
    import sys

    # Decoding value from datapoint 6 which is base64-based (manually activated)
    value_phase_a = state.get('sensor.boiler_lavande_phase_a')
    value_decoded = base64.b64decode(value_phase_a)

    # Calculating Voltage from base64-decoded value
    value_voltage = round(int.from_bytes(value_decoded[0:2], byteorder='big')*0.1, 1)
    state.set('sensor.boiler_lavande_voltage_py', value_voltage)

    # Calculating Current from base64-decoded value
    value_current = round(int.from_bytes(value_decoded[3:5], byteorder='big')*0.001, 2)
    state.set('sensor.boiler_lavande_current_py', value_current)

    # Calculating Power from base64-decoded value
    value_power = round(int.from_bytes(value_decoded[6:8], byteorder='big')*0.001, 3)
    state.set('sensor.boiler_lavande_power_py', value_power)


With this values, i made 3 files in python separated in the same folder where is located the file configuration.yaml:

base64_current.py
base64_power.py
base64_voltage.py

---------------------------------------------------------------------------------------------------
File1 : base64_current.py
---------------------------------------------------------------------------------------------------

```
#!/usr/bin/env python3
import base64
import sys

data = base64.b64decode(sys.argv[1])
if len(sys.argv) >= 3:
    length = int(sys.argv[2])
    data = data[:length]
value_current = round(int.from_bytes(data[3:5], byteorder='big')*0.001, 3)

print(value_current)
```

----------------------------------------------------------------------------------------------------
File 2: base64_power.py
----------------------------------------------------------------------------------------------------

```
#!/usr/bin/env python3
import base64
import sys

data = base64.b64decode(sys.argv[1])
if len(sys.argv) >= 3:
    length = int(sys.argv[2])
    data = data[:length]
value_power = round(int.from_bytes(data[6:8], byteorder='big')*1, 3)

print(value_power)
```

---------------------------------------------------------------------------------------------------
File 3: base64_voltage.py
---------------------------------------------------------------------------------------------------

```
#!/usr/bin/env python3
import base64
import sys

data = base64.b64decode(sys.argv[1])
if len(sys.argv) >= 3:
    length = int(sys.argv[2])
    data = data[:length]
value_voltage = round(int.from_bytes(data[0:2], byteorder='big')*0.1, 1)

print(value_voltage)
```

---------------------------------------------------------------------------------------------------
Then i have configured the sensors in file configuration.yaml


```
command_line:

  - sensor:
      name: Tensão da fase A
      unique_id: tensao_da_fase_a
      state_class: measurement
      command: "python3 /config/base64_voltage.py {{ states('sensor.phase_a') }}"
      value_template: "{{ value }}"
      unit_of_measurement: "V"
      device_class: voltage
      scan_interval: 1

  - sensor:
      name: Corrente da fase A
      unique_id: corrente_da_fase_a
      command: "python3 /config/base64_current.py {{ states('sensor.phase_a') }}"
      value_template: "{{ value }}"
      unit_of_measurement: "A"
      state_class: measurement
      device_class: current
      scan_interval: 1

  - sensor:
      name: Potência da fase A
      state_class: measurement
      unique_id: potencia_da_fase_a
      command: "python3 /config/base64_power.py {{ states('sensor.phase_a') }}"
      value_template: "{{ value }}"
      unit_of_measurement: "W"
      device_class: power
      scan_interval: 1

```
----------------------------------------------------------------------------------------------------
The name "sensor.phase_a" is from the dp_id 6 that i configured in my device with localtuya that have the values CWYAAAAAAAA= x, you can change to the name that you have, the in lovelace you can add the card with the unique_id configured in configuration.yaml, in my case i put, tensao_da_fase_a, corrente_da_fase_a, and the last, potencia_da_fase_a, because im Portuguese. 

I hope that this cold help someone, i have the circuit breaker: 


(https://www.amazon.es/dp/B0C4F3D3XL/ref=asc_df_B0C4F3D3XL/?tag=ptgogshpadde-21&linkCode=df0&hvadid=641091759744&hvpos=&hvnetw=g&hvrand=16816607827984185062&hvpone=&hvptwo=&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=20871&hvtargid=pla-2196315584156&language=pt_PT&mcid=e3317047bcc032cb95800a4d66ca26ca&th=1)

**Comment by lfarkas** [comment]
Any chance that this will be included in the official upstream version with need to manually hack the confirmation files?

**Comment by andrus2049** [comment]
> Then i have configured the sensors in file configuration.yaml
> 
> ```
> command_line:
> 
>   - sensor:
>       name: Tensão da fase A
>       unique_id: tensao_da_fase_a
>       state_class: measurement
>       command: "python3 /config/base64_voltage.py {{ states('sensor.phase_a') }}"
>       value_template: "{{ value }}"
>       unit_of_measurement: "V"
>       device_class: voltage
>       scan_interval: 1
> 
>   - sensor:
>       name: Corrente da fase A
>       unique_id: corrente_da_fase_a
>       command: "python3 /config/base64_current.py {{ states('sensor.phase_a') }}"
>       value_template: "{{ value }}"
>       unit_of_measurement: "A"
>       state_class: measurement
>       device_class: current
>       scan_interval: 1
> 
>   - sensor:
>       name: Potência da fase A
>       state_class: measurement
>       unique_id: potencia_da_fase_a
>       command: "python3 /config/base64_power.py {{ states('sensor.phase_a') }}"
>       value_template: "{{ value }}"
>       unit_of_measurement: "W"
>       device_class: power
>       scan_interval: 1
> ```
As you said, there is no need to modify the core.entity_registry file, as this is a dangerous task...
But you could simply create 3 template sensors, which are faster and simpler than command-line sensors, and still can have all the required device classes and units.
Everything works perfectly with the original base64_int.py, of course using the same "whatever you like" sensor names in both places.
```
- platform: template
  sensors:
    whatever_current_py:
      friendly_name: "whatever phase A - current"
      value_template: 0
      device_class: current
      unit_of_measurement: A
      unique_id: "whatever - phase A - current"

    whatever _voltage_py:
      friendly_name: "whatever  - phase A - voltage"
      value_template: 0
      device_class: voltage
      unit_of_measurement: V
      unique_id: "whatever  - phase A - voltage"

    whatever _power_py:
      friendly_name: "whatever  - phase A - power"
      value_template: 0
      device_class: power
      unit_of_measurement: W
      unique_id: "whatever  - phase A - power"

```

**Comment by Tigranik90** [comment]
I decided to take the easy way. I created a function in NodeRed and 3 sensors.
![Image](https://github.com/user-attachments/assets/1142d4ab-4441-4c65-82a3-9fa6095b772b)

**function code:**
```
let base64data = msg.payload;
let buffer = Buffer.from(base64data, 'base64');
let voltage = buffer.readUInt16BE(0) / 10.0;
let electriccurrent = Buffer.concat([Buffer.from([0x00]), buffer.slice(2, 5)]).readUInt32BE(0) / 1000.0;
let power = Buffer.concat([Buffer.from([0x00]), buffer.slice(5, 8)]).readUInt32BE(0) / 1000.0;
msg.voltage = voltage;
msg.electriccurrent = electriccurrent;
msg.power = power;
return msg;
```

 As a result, I got two devices, but by naming them the same, I added all the sensors to one panel.
![Image](https://github.com/user-attachments/assets/c8b53d94-7d44-4282-9fc7-41950ede4864)

_p.s. In order to get the data in Raw format, I had to add DPS 6 to the Tuya device configuration._
![Image](https://github.com/user-attachments/assets/815c35f4-cd77-4556-afd5-4991748aadf9)
